Framework/API/inc/MantidAPI/IFunction.h

```diff
@@ -254,6 +295,28 @@ class MANTID_API_DLL IFunction {
     /// Apply a lambda visitor
     template <typename... Ts> void apply(AttributeLambdaVisitor<Ts...> &v) { boost::apply_visitor(v, m_data); }
 
+    /// Set Attribute Name
+    void setName(const std::string &name) const;
```

Comment:
This is implemented this way as the existing `declareAttribute(const std::string &name, const API::IFunction::Attribute &defaultValue)` takes a reference to a `const API::IFunction::Attribute` as an argument. 

We would have to make this non-`const` in this function, and in related functions throughout derived classes to simply remove const from `setName` and `mutable` from `m_name`.

Ideally the attribute would be maintained as a `const`, though I've introduced the `m_validator` member variable to `Attributes` which is not set during attribute initialisation but via `setAttribute` and therefore needs to be mutable.

Ideally this would be set during initialisation and then would not need to be `mutable`, however I was (a) aiming for the `declareAttribute("Unit", Attribute("K"), StringListValidator({"K", "meV"}))` syntax as prescribed in the issue, rather than `declareAttribute("Unit", Attribute("K", StringListValidator({"K", "meV"}))`, and (b) wanting to leave the attribute implementation largely alone as unsure about knock-on consequences.

Now I'm more familiar with this bit of the code/the python interface I think the later would probably work too and be fairly undisruptive, though not sure if its worth the effort. I'll take a look.